<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />
<script src="./bower_components/jquery/dist/jquery.min.js"></script>
<script type="text/javascript">
$(function() {

$.ajax({
    //異常系を修正する場合はわざとurlを崩したり、エラーを出すなど
    url:'/library/ajaxpost.php',
    type:'POST',
    data:{
        'user_id':'aaaa',
        'password':'bbbb'
    }
}).done((data) => {
    //成功時
    console.log(data)
    console.log('ajax success')
}).fail((data) => {
    //失敗時
    console.log('ajax fail 404, 500エラーなど');
}).always((data) => {
    //try-catchのfinallyに近い
    //ここは常に通る
    console.log('ajax always')
})


function getPref() {
    return new Promise((resolve, reject) => {
        $.ajax({
          url:'/library/ajaxpost.php',
          type:'POST',
          data:{
            'user_id':'aaaa',
            'password':'bbbb'
          }
        }).done((data) => {
          resolve(data)
        }).fail((data) => {
          reject(data)
        })
    })
}

console.log('promise');

getPref()
.then((data) => {
    console.log("promise success")
    console.log(data)
}).catch((err) => {
    console.log("promise error 404 500 error")
    console.log(err)
}).finally(()=>{
    console.log("promise finally")
});

//promiseを返す
//https://qiita.com/suin/items/97041d3e0691c12f4974
async function getPrefProcess() {
    //読んで字のごとくawait結果をまつ
    try{
        console.log("async_await1 start")
        const result =  await getPref();
        console.log(result)
    } catch(error) {
        console.log('async_awit1 error');
        console.log(error)
    } finally {
        console.info('async_await1 finally');
    }
}

getPrefProcess()


//async_awit promiseを返す throw
async function getPrefProcess2() {

    try{
        console.log("async_await2 start")
        const result =  await getPref();
        console.log(result)
    } catch(error) {
        console.log('async_awit2 error throw');
        throw error;
    }
}


getPrefProcess2()
.then(()=>{
    console.log("async_awit wrapper2 then")
}).catch((err)=>{
    console.log("async_awit wrapper2 catch")
    console.log(err)
}).finally(()=>{
    console.log("async_awit wrapper2 finally")
})

})
</script>
<body>
    <div id="pref_area"></div>
    <div id="city_area"></div>
</body>
</html>
